crates/router/src/core/payments/payment_methods.rs

```diff
@@ -0,0 +1,270 @@
+struct PaymentMethodsEnabledWithRequiredFieldsAndSurcharge {
+    required_field:
+        Option<std::collections::HashMap<String, api_models::payment_methods::RequiredFieldInfo>>,
+    payment_method_subtype: common_enums::PaymentMethodType,
+    surcharge: Option<api_models::payment_methods::SurchargeDetailsResponse>,
+}
+
+/// Container to hold the filtered payment methods enabled with required fields and surcharge
+struct PaymentMethodsEnabledWithRequiredFieldsAndSurchargeContainer {
+    payment_methods_enabled: Vec<PaymentMethodsEnabledWithRequiredFieldsAndSurcharge>,
+}
+
+impl PaymentMethodsEnabledWithRequiredFieldsAndSurchargeContainer {
+    fn generate_response(self) -> api_models::payments::PaymentMethodListResponseForPayments {
+        let response_payment_methods = self
+            .payment_methods_enabled
+            .into_iter()
+            .map(|payment_methods_enabled| {
+                api_models::payment_methods::ResponsePaymentMethodTypes {
+                    payment_method_subtype: payment_methods_enabled.payment_method_subtype,
+                    required_fields: payment_methods_enabled.required_field,
+                    surcharge_details: payment_methods_enabled.surcharge,
+                    card_networks: None,
+                    bank_names: None,
+                }
+            })
+            .collect();
+
+        api_models::payments::PaymentMethodListResponseForPayments {
+            payment_methods_enabled: response_payment_methods,
+            customer_payment_methods: Vec::new(),
+        }
+    }
+}
+
+impl PaymentMethodsEnabledWithRequiredFieldsContainer {
+    fn perform_surcharge_calculation(
+        self,
+    ) -> PaymentMethodsEnabledWithRequiredFieldsAndSurchargeContainer {
+        let details_with_surcharge = self
+            .payment_methods_enabled
+            .into_iter()
+            .map(
+                |payment_methods_enabled| PaymentMethodsEnabledWithRequiredFieldsAndSurcharge {
+                    required_field: payment_methods_enabled.required_field,
+                    payment_method_subtype: payment_methods_enabled.payment_method_subtype,
+                    surcharge: None,
+                },
+            )
+            .collect();
+
+        PaymentMethodsEnabledWithRequiredFieldsAndSurchargeContainer {
+            payment_methods_enabled: details_with_surcharge,
+        }
+    }
+}
+
+/// Holds the payment methods enabled for a connector
+struct PaymentMethodsEnabled {
+    payment_methods_enabled: Vec<PaymentMethodsEnabledForConnector>,
+}
+
+impl PaymentMethodsEnabled {
+    /// This functions flattens the payment methods enabled from the connector accounts
+    /// Retains the connector name and payment method in every flattened element
+    fn from_payment_connectors_list(
+        payment_connectors: Vec<domain::MerchantConnectorAccount>,
+    ) -> Self {
+        let payment_methods_enabled_flattened_with_connector = payment_connectors
```

Comment:
This function can be moved to domain::MerchantConnectorAccounts, as flatten_payment_method_sub_types.